pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java

```diff
@@ -43,4 +48,61 @@ public static TransferableBlock getErrorTransferableBlock(Map<Integer, String> e
   public static boolean isEndOfStream(TransferableBlock transferableBlock) {
     return transferableBlock.isEndOfStreamBlock();
   }
+
+  /**
+   *  Split a block into multiple block so that each block size is within maxBlockSize.
+   *  Currently, we only support split for row type dataBlock.
+   *  For columnar data block, we return the original data block.
+   *  Metadata data block split is not supported.
+   *
+   *  When row size is greater than maxBlockSize, we pack each row as a separate block.
+   */
+  public static List<BaseDataBlock> getDataBlockChunks(BaseDataBlock block, BaseDataBlock.Type type, int maxBlockSize)
+      throws IOException {
+    List<BaseDataBlock> blockChunks = new ArrayList<>();
+    switch (type) {
+      // TODO: Add support for columnar data block.
+      case COLUMNAR:
+        blockChunks.add(block);
+        return blockChunks;
+      case METADATA:
+        throw new UnsupportedOperationException("splitBlock is not supported for metdata block.");
+      case ROW:
+        break;
+      default:
+        throw new UnsupportedOperationException("splitBlock is not supported for type:" + type);
+    }
+    // TODO: Store row size in bytes inside data block.
+    // TODO: Calculate dictionary and var bytes size as well.
+    DataSchema dataSchema = block.getDataSchema();
+    int[] columnOffsets = new int[dataSchema.size()];
+    int rowSizeInBytes = DataBlockUtils.computeColumnOffsets(dataSchema, columnOffsets);
+    List<Object[]> chunk = new ArrayList<>();
+    int numRowsPerChunk = maxBlockSize / rowSizeInBytes;
+    long numRows = 0;
+    List<Object[]> container = DataBlockUtils.extractRows(block);
```

Comment:
this will deserialize the entire block just to re-serialize it again later when we `buildFromRows` - which I suspect will be a significant performance regression for large blocks. is there any way to do this split without going through the serde twice?

Reply:
this is the first step. we will optimized this one out later